### PR TITLE
Audio View: go to measure

### PR DESCRIPTION
**Pull request by nittka**
Proof of concept: add UI control to let the user enter a measure number.
Based on beats per measure and beats per second (extracted from midi),
the time to go to is calculated.
Of course this works only if the tempo or beats per measure do not
change.

### CODE REVIEW

**Comment by nittka** [comment]
I am thinking of including this feature into the refactored audio control.
In order to create reliable results; I would require the lilypond file to contain a comment like 
`%elysium-midi: 6/4, partial 9/16`
Using such a comment, the user indicates that there are no measure or tempo changes, so that the current measure can be calculated. In that case, an additional text field would be shown in the button row, showing the current measure and allowing to enter the measure to go to.

**Comment by nittka** [comment]
I have updated the code to the new audio control. Rather than having three fields, there is now one textfield - initially empty.
You can enter a measure number here and the approximate time is calculated. Beats per minute are extracted from the midi. Beats per measure are assumed to be 4 by default. The field format allows to override the defaults:
* "6" - go to the 6th measure
* "9:3" - go the 9th measure assuming 3 beats per measure
* "17:5:113" - go to the 17th measure assuming 5 beats per measure and 113 beats per minute

As the field is initially empty and the measure number is updated only if the field is not empty, the visual impact is quite small.
Currently the overriding of the defaults is activated only when pressing return. This could be extracted to a separate method, so that the update is done on every keystroke in the field. (Note that the measure number is updated not only when playing but also when using the slider)

**Comment by thSoft** [comment]
Could you calculate the time from the tempo signature and tempo change events present in the MIDI file? IMHO this is unfortunately necessary because these can change over time.

**Comment by nittka** [comment]
I have not worked with midi events, so far. It may be possible (if quite expensive for this feature, as the events would have to be tracked permanently - during playback, when editing the field, when using the slider) to extract *some* more information improving the result, but it will never be perfect. The information about partial measures, measure changes, (repetitions...) in a piece will not be present in the midi, rendering even tracking tempo changes quite useless.

I think, this is an additional little tool. It is useful in some situations, and can easily be igored otherwise. As I said, due to the default (unset), the visual impact is quite small (and could be reduced further by changing the background color of the text field).

https://www.dummies.com/art-center/music/recording-music/midi-time-code-and-midi-clock-explained/

http://midi.teragonaudio.com/tech/midifile/time.htm

http://www.harfesoft.de/aixphysik/sound/midi/pages/miditmcn.html

**Comment by nittka** [comment]
As far as I can see, there is no way to extract information about partial measures from the midi file. However, there are events for measure changes (and tempo changes). As you can get (and set) the tick position and the microseconds position from the sequencer, the simplest way would be to ignore tempo change events. Using the measure events (3/4, 6/8,..), you calculate the tick of the start of each measure, set the tick position in the sequencer and ask for the time in microseconds.
That way you can create a Map measure number -> time. For displaying the measure number/going to a position, you then consult this map.
Tempo changes should be handled implicitly, as the sequencer should return the correct microposition change for a tick.

**Comment by thSoft** [comment]
Yes, that sounds like a great plan, thanks for the investigation! It would be great if you took the effort to implement it.


________________________________
Feladó: Alexander Nittka <notifications@github.com>
Elküldve: Saturday, February 27, 2021 1:38:55 PM
Címzett: thSoft/eclipse-commons <eclipse-commons@noreply.github.com>
Másolatot kap: Dénes Harmath <harmathdenes@gmail.com>; Review requested <review_requested@noreply.github.com>
Tárgy: Re: [thSoft/eclipse-commons] Audio View: go to measure (#53)


As far as I can see, there is no way to extract information about partial measures from the midi file. However, there are events for measure changes (and tempo changes). As you can get (and set) the tick position and the microseconds position from the sequencer, the simplest way would be to ignore tempo change events. Using the measure events (3/4, 6/8,..), you calculate the tick of the start of each measure, set the tick position in the sequencer and ask for the time in microseconds.
That way you can create a Map measure number -> time. For displaying the measure number/going to a position, you then consult this map.
Tempo changes should be handled implicitly, as the sequencer should return the correct microposition change for a tick.

**Comment by nittka** [comment]
I pushed a first prototype. Tempo and time changes should be handled, already.

Partial measures will be dealt with in a further commit.

**Comment by nittka** [comment]
Now partials are handled as well. \unfoldRepeats in midi cannot be handled, though. (Before a merge, I would like to squash and cleanup.)

**Comment by thSoft** [comment]
Thank you very much, the implementation works great as far as I could test! I'd like to have a few UX proposals:
* Add labels before the text fields to make their meanings clear: *Measure:* and *Time in measure:*
* Always update measure number during playback (this makes its meaning even more clear)

**Comment by nittka** [comment]
I added labels and made minimal adjustments to always show the measure number. With all labels and text fields in a row, the view gets quite wide. So I tried to arrange the new fields to allow quite a compact presentation.
If you are happy with the result, I would clean up the code, possibly moving some of the code so most of the midi stuff is contained in one class and squash everything in one commit.

Please check the tooltip texts. I am not sure if the text for partial is quite clear.

File: org.eclipse.ui.views.midi/src/org/eclipse/ui/views/midi/MidiPlaybackControl.java [hunk 1]
```diff
@@ -140,6 +150,68 @@ public void widgetSelected(SelectionEvent e) {
...
+		mLayout.marginBottom=-5;
+		mLayout.marginTop=-5;
+		mParent.setLayout(mLayout);
+
+		GridData measureData=new GridData();
+		measureData.widthHint=35;
+		measureData.grabExcessHorizontalSpace=false;
+
+		partialLabel= new Label(mParent, SWT.NONE);
+		partialLabel.setText("Partial:");
```
Comment (thSoft) [inline org.eclipse.ui.views.midi/src/org/eclipse/ui/views/midi/MidiPlaybackControl.java]:
```suggestion
		partialLabel.setText("Upbeat:");
```

File: org.eclipse.ui.views.midi/src/org/eclipse/ui/views/midi/MidiPlaybackControl.java [hunk 2]
```diff
@@ -140,6 +150,68 @@ public void widgetSelected(SelectionEvent e) {
...
+		partial = new Text(mParent, SWT.NONE);
+		partial.setText("");
+		partial.setTextLimit(6);
+		partial.setToolTipText("If the score starts with a partial measure,\nplease provide that information!\nFormat examples '1/4', '3/8'");
```
Comment (thSoft) [inline org.eclipse.ui.views.midi/src/org/eclipse/ui/views/midi/MidiPlaybackControl.java]:
```suggestion
		partial.setToolTipText("If the score starts with a partial measure,\nplease provide its duration!\nFormat examples '1/4', '3/8'");
```

File: org.eclipse.ui.views.midi/src/org/eclipse/ui/views/midi/MidiPlaybackControl.java [hunk 3]
```diff
@@ -140,6 +150,68 @@ public void widgetSelected(SelectionEvent e) {
...
+				updateUpdateMeasureActivation();
+			}
+		});
+		partial.setLayoutData(measureData);
+		
+		new Label(mParent, SWT.NONE).setText("Measure:");
+		measure = new Text(mParent, SWT.NONE);
+		measure.setText("");
+		measure.setTextLimit(4);
+		measure.setToolTipText("Enter measure number and press return to go to approximate time!");
```
Comment (thSoft) [inline org.eclipse.ui.views.midi/src/org/eclipse/ui/views/midi/MidiPlaybackControl.java]:
```suggestion
		measure.setToolTipText("Enter measure number and press Return to go to approximate time!");
```

**Review by thSoft — CHANGES_REQUESTED** [review]
Sorry for the delay!
* Don't hide the "partial" field during playback, only disable it.
* Navigate to the desired point of time also when pressing Enter in the "partial" field.

**Comment by nittka** [comment]
I have updated the code according to your suggestions (thanks for the help on wording!). I also cleaned up the code - extracting most of the measure control parts into an inner class and squashed everything into one commit.
(If the midi analysis fails (basically if an exception occurs) all the measure controls are hidden. I tested the code by simply returning false in the analyzers calculateMeasureTimes method in debug mode.)

If you have no further change requests, I am happy with the current state.

**Review by thSoft — APPROVED** [review]

**Pull request merged**